### ISSUE TITLE
Recommended way to reset ajaxHeaders

### ISSUE THREAD

**Issue by oceanful** [openseadragon/openseadragon#1748]
Our image requests are authorized with an  HTTP header that must be reset once an hour.

The way we've partially made this work is by periodically resetting the `ajaxHeaders` property on the `Viewer` and all `TiledImage`'s returned from `Viewer.world.getItemAt()`. 

The problem is that the new `ajaxHeaders` do not take effect for all tile requests. For some tiles , requests are made  with old `ajaxHeaders`, sometimes multiple generations back. (It's often for tiles that have been previously requested, but it seems to be somewhat random). This happens even if we call `TiledImage.reset()` to clear the internal `TileCache` after resetting the `ajaxHeaders`.

1. First of all, is our method of resetting `ajaxHeaders` the best way to achieve our goal, or is there a better approach?
2. If our approach is reasonable, how do we fix the problem of stale `ajaxHeaders`?

Thank you!

**Comment by iangilman** [comment]
Interesting! Yes, it looks like the `ajaxHeaders` get copied into many parts of the system, starting at the Viewer but going down into the TileSource, TiledImage, Tile, ImageLoader, and the ReferenceStrip.

I think a robust solution would be for Viewer to have a `setAjaxHeaders` function that makes sure to push the updated headers into all of its members that need it. For instance, it would tell its `world` to tell of its `_items` (which are all TiledImages). It should also tell its `tileCache` to tell of its `_tilesLoaded`, etc.

Would you be up for writing such a patch?

**Comment by oceanful** [comment]
Thanks Ian!

For now, we are working around this by (1) resetting `viewer.ajaxHeaders`, (2) adding new tiled images under the old ones, and (3) removing the old ones when the new ones are fully loaded.

Unfortunately we don't have the bandwidth to make the change you proposed. Hopefully someone will at some point!

**Comment by iangilman** [comment]
Sounds fair. I'm glad you have something that works!

**Comment by uschmidt83** [comment]
> For some tiles , requests are made with old ajaxHeaders, sometimes multiple generations back. (It's often for tiles that have been previously requested, but it seems to be somewhat random).

FYI, I just had the same problem and it seems that it can be solved like this:

```javascript
for (let i = 0; i < viewer.world.getItemCount(); i++) {
    const image = viewer.world.getItemAt(i);

    for (const level of Object.values(image.tilesMatrix))
      for (const row of Object.values(level))
        for (const tile of Object.values(row)) tile.ajaxHeaders = newAjaxHeaders;
  }
}
```

**Comment by iangilman** [comment]
@uschmidt83 Lovely! This seems like it would be a good start to a patch to add a `setAjaxHeaders` function. Would you be interested in making a PR for that? If not, no worries... Thank you for sharing the code!

**Comment by uschmidt83** [comment]
Hi @iangilman, thanks for the reply.

I could in principle make a PR, but I'm not sure I understand the codebase and different uses cases well enough. I've looked at all mentions of `ajaxHeaders` that I could find, and that's my current understanding:

- The main [options](http://openseadragon.github.io/docs/OpenSeadragon.html#.Options) have an `ajaxHeaders` field, which is stored by the Viewer in `Viewer.ajaxHeaders`.
- [`Viewer.addTiledImage(options)`](https://github.com/openseadragon/openseadragon/blob/b1274515aa24a1d85c74cd0017673d7a1884ac85/src/viewer.js#L1418) takes `options.ajaxHeaders`, which will be merged with the main `Viewer.ajaxHeaders` to determine `TiledImage.ajaxHeaders` (for the created `TiledImage`). Also, the boolean flag `options.loadTilesWithAjax` has to be enabled here or already in the main options.
Furthermore, one can pass a function [`options.tileSource.getTileAjaxHeaders(level, x, y)`](https://github.com/openseadragon/openseadragon/blob/b1274515aa24a1d85c74cd0017673d7a1884ac85/src/tilesource.js#L672) to "override headers specified at the Viewer or TiledImage level." as stated in the documentation.
- [`TiledImage._getTile(...)`](https://github.com/openseadragon/openseadragon/blob/b1274515aa24a1d85c74cd0017673d7a1884ac85/src/tiledimage.js#L1470) creates new `Tile`s with `Tile.ajaxHeaders` determined by mixing `TiledImage.ajaxHeaders` with the returned headers from calling `tileSource.getTileAjaxHeaders(...)`. A given `Tile` is created only once and stored in `TiledImage.tilesMatrix[level][x][y]`.
- [`TiledImage._loadTile(tile, ...)`](https://github.com/openseadragon/openseadragon/blob/b1274515aa24a1d85c74cd0017673d7a1884ac85/src/tiledimage.js#L1568) calls `ImageLoader.addJob(options)` with `options.ajaxHeaders = tile.ajaxHeaders`.

If the values in the `ajaxHeaders` need to change (e.g. new Bearer token), the main problem is that all `Tile`s in `TiledImage.tilesMatrix[level][x][y]` store their own `Tile.ajaxHeaders` property, which need to be updated for `TiledImage._loadTile(tile, ...)` to work correctly again.
Of course, `Viewer.ajaxHeaders`/`TiledImage.ajaxHeaders` must be updated as well, unless one is only using `tileSource.getTileAjaxHeaders(...)` which can dynamically always return the correct headers, for new `Tile`s that have never been requested.

Furthermore, I think one should update all `Tile.cacheKey` (default value is `tileSource.getTileHashKey(level, x, y, url, ajaxHeaders, postData)`), assuming the images are not different when the `ajaxHeaders` change. But one could also implement `tileSource.getTileHashKey(...)` to ignore `ajaxHeaders`, but this requires a custom tile source.

---

Now, one could add a `setAjaxHeaders` function to `Viewer` and `TiledImage`, which could propagate the changes to all `Tile`s in `TiledImage.tilesMatrix`. The problem I see with this is the merging of different `ajaxHeaders`, because only the "merged" headers are stored and not the original headers before merging. Also, the header entries from `options.tileSource.getTileAjaxHeaders(level, x, y)` take precedent over anything in `TiledImage.ajaxHeaders` and would require re-evaluation of `getTileAjaxHeaders` for all tiles?

I think there are too many different ways how `ajaxHeaders` can be used and supporting all different variants seems complicated and confusing to me. But maybe I'm misunderstanding something here and there's a clear way how this could be implemented.

**Comment by iangilman** [comment]
@uschmidt83 Thank you for mapping out how the ajaxHeaders propagate through the system! 

It seems to me a `Viewer.setAjaxHeaders` would need to go through that same process again: 

* Set `Viewer.ajaxHeaders`.
* Go through all of the TiledImages and set their ajaxHeaders, once again merging with whatever headers they were given by the `addTiledImage` (which we'll need to store separately now so we can come back to it). 
* Go through all of the Tiles and update their `ajaxHeaders` by combining the new TiledImage headers with a fresh call to get the TileSource headers. 

I feel like we don't necessarily need to update any ImageLoader jobs that are currently in flight, but maybe that too? 

I don't know that we need to update the Tile cacheKeys (assuming you don't want them to count as an entirely new tile)... The Tiles get their key when they are created and mostly they just need to be unique. @Aiosa might have other thoughts on that front. 

We can keep talking through it if you want to pursue the patch, but no worries if not. Even what you've done so far is very helpful for whoever might make the change!

**Comment by uschmidt83** [comment]
I've created a PR #2346 to implement this.

**Comment by iangilman** [comment]
Fixed by #2346.
